Engineering/useEffect double calls in React 18.md

```diff
@@ -0,0 +1,78 @@
+---
+tags: engineering, react 18, react hooks, useEffect
+author: T.Hoang Nam (Nigel)
+---
+
+### useEffect double calls in React 18
+
+In the React 18 version, the useEffect hook has been updated to called twice compare to only one in the older version in StrictMode.
+
+**What are the changes and why?**
+- Beginning with React 18, when in development mode, the components will be mounted, unmounted, and then mounted once again in StrictMode.
+- This was introduced so that in the future, when React decides to offer a feature that allows it to add or delete an area of the UI while still maintaining the state, this will help it do so. For instance, when moving between tabs, maintaining the state of the preceding tab might assist avoid the execution of effects such as API calls that are not essential.
```

Comment:
I think the paragraph doesn't quite clearly explain the reason for this new behavior, especially, in your example. Below's the takeaway from the [React Docs](https://reactjs.org/blog/2022/03/29/react-v18.html#new-strict-mode-behaviors) that might be a reference if you want to update the section:

> In the future, we’d like to add a feature that allows React to add and remove sections of the UI while preserving state. For example, when a user tabs away from a screen and back, React should be able to immediately show the previous screen. To do this, React would unmount and remount trees using the same component state as before.

> This feature will give React apps better performance out-of-the-box, but requires components to be resilient to effects being mounted and destroyed multiple times. Most effects will work without any changes, but some effects assume they are only mounted or destroyed once.

> To help surface these issues, React 18 introduces a new development-only check to Strict Mode. This new check will automatically unmount and remount every component, whenever a component mounts for the first time, restoring the previous state on the second mount.